upgrade-2.x.sh

```diff
@@ -0,0 +1,204 @@
+            shift
+            ;;
+        *)
+            loglevel=LOG
+            ;;
+    esac
+    endtime=$(date +%s)
+    if [ "z$LOG" == "zyes" ] && [ -n "$LOGFILE" ]; then
+        printf "[%09d%s%s\n" "$((endtime - starttime))" "][$loglevel]" "$1" | tee -a "$LOGFILE"
+    else
+        printf "[%09d%s%s\n" "$((endtime - starttime))" "][$loglevel]" "$1"
+    fi
+    if [ "$loglevel" == "ERROR" ]; then
+        /usr/bin/resin-device-progress --percentage 100 --state "ResinOS: Update failed."
+        exit 1
+    fi
+}
+
+# Test if a version is greater than another
+function version_gt() {
+    test "$(echo "$@" | tr " " "\n" | sort -V | head -n 1)" != "$1"
+}
+
+# Log timer
+starttime=$(date +%s)
+
+# Check board support
+case $SLUG in
+    beaglebone*)
+        binary_type=arm
+        ;;
+    raspberry*)
+        binary_type=arm
+        ;;
+    *)
+        log ERROR "Unsupported board type $SLUG."
+esac
+
+if [ -n "$target_version" ]; then
+    target_version=$(echo "$target_version" | tr + _)
+    case $target_version in
+        2.*)
+            log "Target OS version \"$target_version\" OK."
+            ;;
+        *)
+            log ERROR "Target OS version \"$target_version\" not supported."
+            ;;
+    esac
+else
+    log ERROR "No target OS version specified."
+fi
+
+# Check host OS version
+case $VERSION in
+    2.*)
+        log "Host OS version \"$VERSION\" OK."
+        ;;
+    *)
+        log ERROR "Host OS version \"$VERSION\" not supported."
+        ;;
+esac
+
+# Check if we need to install some more extra tools
+if ! version_gt "$VERSION" "$preferred_hostos_version" &&
+    ! [ "$VERSION" == "$preferred_hostos_version" ]; then
+    log "Host OS version $VERSION is less than $preferred_hostos_version, installing tools..."
+    tools_path=/tmp/upgrade_tools
+    tools_binaries="tar"
+    mkdir -p $tools_path
+    export PATH=$tools_path:$PATH
+    case $binary_type in
+        arm)
+            download_uri=https://github.com/resin-os/resinhup/raw/onetotwo/upgrade-binaries/$binary_type
+            for binary in $tools_binaries; do
+                log "Installing $binary..."
+                curl -s -L -o $tools_path/$binary $download_uri/$binary
+                chmod 755 $tools_path/$binary
+            done
+            ;;
+        *)
+            log ERROR "Binary type $binary_type not supported."
+            ;;
+    esac
+fi
+
+# Find which partition is / and which we should write the update to
+root_part=$(findmnt -n --raw --evaluate --output=source /)
+case $root_part in
+    *p2)
+        root_dev=${root_part%p2}
+        update_part=${root_dev}p3
+        update_part_no=3
+        update_label=resin-rootB
+        ;;
+    *p3)
+        root_dev=${root_part%p3}
+        update_part=${root_dev}p2
+        update_part_no=2
+        update_label=resin-rootA
+        ;;
+    *)
+        log ERROR "Unknown root partition \"$root_part\"."
+esac
+
+# Stop docker containers
+log "Stopping all containers..."
```

Comment:
If we don't do that then something (I guess the supervisor) cleans up containers which I think can make the export of the container fail below.